File: lib/queue/message-sync-queue.ts
/**
 * Message Sync Queue
 * 
 * BullMQ queue for syncing LinkedIn messages.
 * Runs as a repeating job every 3 minutes.
 */

import { Queue } from 'bullmq'
import { Redis } from 'ioredis'

const connection = new Redis(process.env.REDIS_URL || 'redis://localhost:6379', {
  maxRetriesPerRequest: null,
  enableReadyCheck: false,
  retryStrategy: (times) => {
    if (times > 20) return null
    return Math.min(times * 500, 5000)
  },
})

connection.on('error', (err) => {
  console.error('[MessageSyncQueue] Redis error:', err.message)
})

export const MESSAGE_SYNC_QUEUE = 'message-sync'

export const messageSyncQueue = new Queue(MESSAGE_SYNC_QUEUE, {
  connection: connection as any,
  defaultJobOptions: {
    removeOnComplete: 10,
    removeOnFail: 20,
  },
})

export interface MessageSyncJobData {
  type: 'sync-all' | 'sync-account' | 'send-reply'
  linkedinAccountId?: string
  userId?: string
  threadId?: string
  messageText?: string
  conversationId?: string
}

/**
 * Schedule a repeating sync job (call once on boot)
 */
export async function scheduleMessageSync(intervalMinutes: number = 3) {
  // Remove any existing repeating jobs to avoid duplicates
  const existing = await messageSyncQueue.getRepeatableJobs()
  for (const job of existing) {
    await messageSyncQueue.removeRepeatableByKey(job.key)
  }

  // Add a repeating sync job
  await messageSyncQueue.add(
    'sync-all-accounts',
    { type: 'sync-all' as const },
    {
      repeat: {
        every: intervalMinutes * 60 * 1000, // every N minutes
      },
      removeOnComplete: 5,
      removeOnFail: 10,
    },
  )

  console.log(`[MessageSync] Scheduled repeating sync every ${intervalMinutes} minutes`)
}

/**
 * Queue a single reply to be sent
 */
export async function queueReply(data: {
  linkedinAccountId: string
  threadId: string
  messageText: string
  conversationId: string
}) {
  await messageSyncQueue.add('send-reply', {
    type: 'send-reply' as const,
    ...data,
  }, {
    removeOnComplete: 20,
    removeOnFail: 5,
    attempts: 2,
    backoff: { type: 'exponential', delay: 10000 },
  })
}

File: backend/internal/queue/message_sync_queue.go
package queue

import (
	"log"
	"time"
)

// ──────────────────────────────────────────────────────────────────────────────
// Message Sync Queue Helpers
//
// Mirrors lib/queue/message-sync-queue.ts:
//   • scheduleMessageSync(intervalMinutes) — repeating "sync-all" job
//   • queueReply(data) — one-shot "send-reply" job
// ──────────────────────────────────────────────────────────────────────────────

const messageSyncRepeaterKey = "message-sync-repeater"

// MessageSyncQueueHelper provides convenience methods for message-sync jobs.
type MessageSyncQueueHelper struct {
	qm *QueueManager
}

// NewMessageSyncQueueHelper creates a helper bound to the given QueueManager.
func NewMessageSyncQueueHelper(qm *QueueManager) *MessageSyncQueueHelper {
	return &MessageSyncQueueHelper{qm: qm}
}

// ScheduleMessageSync sets up a repeating "sync-all" job at the given interval.
// Calling again replaces the previous schedule (idempotent).
func (h *MessageSyncQueueHelper) ScheduleMessageSync(intervalMinutes int) {
	if intervalMinutes <= 0 {
		intervalMinutes = 3
	}

	interval := time.Duration(intervalMinutes) * time.Minute

	data := MessageSyncJobData{
		Type: "sync-all",
	}

	h.qm.AddRepeatingJob(messageSyncRepeaterKey, QueueMessageSync, "sync-all-accounts", data, interval)
	log.Printf("[MessageSync] Scheduled repeating sync every %d minutes", intervalMinutes)
}

// StopMessageSync removes the repeating sync job.
func (h *MessageSyncQueueHelper) StopMessageSync() {
	h.qm.RemoveRepeatingJob(messageSyncRepeaterKey)
}

// QueueSyncAccount enqueues a one-shot sync for a specific LinkedIn account.
func (h *MessageSyncQueueHelper) QueueSyncAccount(linkedinAccountID, userID string) (*Job, error) {
	data := MessageSyncJobData{
		Type:              "sync-account",
		LinkedInAccountID: linkedinAccountID,
		UserID:            userID,
	}
	return h.qm.Enqueue(QueueMessageSync, "sync-account", data, 0)
}

// QueueReply enqueues a single reply to be sent via the Playwright worker.
func (h *MessageSyncQueueHelper) QueueReply(linkedinAccountID, threadID, messageText, conversationID string) (*Job, error) {
	data := MessageSyncJobData{
		Type:              "send-reply",
		LinkedInAccountID: linkedinAccountID,
		ThreadID:          threadID,
		MessageText:       messageText,
		ConversationID:    conversationID,
	}
	return h.qm.Enqueue(QueueMessageSync, "send-reply", data, 0)
}
